### DISCUSSION TITLE
负载均衡的节点不稳定，如何避免链接中断？

### DISCUSSION THREAD

**Discussion by renheren** [v2fly/v2ray-core#501] [Q&A]
请问如何在几个不那么稳定的节点上做负载均衡，达到更加稳定的连接效果？

我的主要思路如下：出站连接做负载均衡，从一个代理池中随机选择几个节点组成一个socks入站的代理阵列，然后把v2ray的出站连上这个阵列，每隔几秒测试socks代理阵列中的节点是不是有被阻断的，如果有就杀掉并从代理池中剩余节点里面另选一个启动，保证代理池中大部分连接都是畅通的。这样配置的优点是不用经常去手动切换节点，网页浏览体验不错，但对于一些涉及文件下载的体验就不怎么样了，如果下载过程中遇到节点切换，就经常下载失败。

用代理池的目的是希望能够365天保持连接，用过一段时间的节点不好用的话可以简单地通过增加订阅其他服务商的节点来维护，同一时间只在代理池中选几个节点是为了避免这些节点被同时阻断。

1. 作为测试，我用了两个独立的v2ray进程A、B作为socks入站，然后再用另一个v2ray进程C通过负载均衡连接到A和B。
用 curl -x socks5h://ip-of-C:socks-port -LO remote_big_file 测试通过C的socks入站连接的稳定性。
在下载过程中中断A或B中的一个，会出现两种情况，一种是下载继续，另一种是报错 "curl: (18) transfer closed with xxxxx bytes remaining to read"，这应该是下载单个文件时只会通过A或B其中一个的表现。请问如何配置能够达到负载均衡节点不稳定，但负载均衡后的连接相对稳定的效果呢？

不得不感叹v2ray路由配置起来可以很灵活，让我有了折腾的空间。

2. 我试过配置一个"web+tls+websocket"服务器D作为最终出口，到D的连接通过C的负载均衡，下载连接还是会中断。
也试过另加一个E在A、B之后，C之前，E对A、B做负载均衡，然后通过VMESS出站连接到C的入站，也还是不行。
3. 各种测试，只有一个勉强达到我要的效果，就是在C内套D，D内再套一层wireguard，通过wireguard去下载文件可以勉强达到不中断的效果，但还是有几秒速度为零。
4. 丧心病狂的我，又在wireguard内套了一层socks入站代理，结果打断A或B后下载文件速度有1/2几率会变零，A或B重新启动后，速度也不会再上来。

这样一层套一层，配置极为复杂，且不好管理，测试到此为止吧， 目前就打算用方案2保证出口ip固定，并在应用层面做好网络异常处理了。

**Comment by kslr** [comment]
下载文件失败，取决于下载工具和是否支持断点续传吧

得先实现连接热切换

**Reply by renheren** [reply]
经测试，上面测试中可以给“curl"加“-C -"参数手动续传。不过在下载过程中连接中断会带来一些应用层面的麻烦。比如说用Docker Build的时候，一个”apt install xxx"的包连接中断了，就必须要手动重新执行命令，整个层的"apt install"只能从头开始下载，如果网速比较慢而中断比较频繁就会陷入了”无限循环“。

**Reply by kslr** [reply]
这个问题从设计上是可以在core里解决的，你有兴趣在这块工作吗

**Reply by renheren** [reply]
哎呀，只会Python的业余码农表示爱莫能助。
完全不知道怎么实现，也不知道工程量和开销会有多少。
希望有一天可以实现连接热切换。
